aoc/src/main/scala/AoC2015Day07.scala

```diff
@@ -0,0 +1,66 @@
+object AoC2015Day07 extends App:
+    import scala.io.Source
+
+    sealed trait Instruction:
+        def retName: String
+        def evaluate(wiresMap: Map[String, Int]): Option[Int]
+        def getWireValueByName(wiresMap: Map[String, Int], wireName: String): Option[Int] =
+            wiresMap.get(wireName)
+
+    case class litInput(retName: String, input: Int) extends Instruction:
+        def evaluate(wiresMap: Map[String, Int]) =
+            Some(input)
+
+    case class singleInput(retName: String, inputName: String, op: Int => Int) extends Instruction:
+        def evaluate(wiresMap: Map[String, Int]) =
+            getWireValueByName(wiresMap, inputName).map(op)
+
+    case class doubleInput(retName: String, inputName1: String, inputName2: String, op: (Int, Int) => Int) extends Instruction:
+        def evaluate(wiresMap: Map[String, Int]) =
+            getWireValueByName(wiresMap, inputName1)
+            .flatMap(x => getWireValueByName(wiresMap, inputName2).map(op.curried(x)))
+
+    def solve(forWire: String, instructions: List[Instruction], wireOverrides: Option[Map[String, Int]] = None): Int =
```

Comment:
im almost okay with the total solutions, but for the `Option[Map[String,Int]]` - it would be a good exercise to have `wireOverrides` be of just type `Map[String,Int]` - note that an empty map can be used to signal that there are overrides ;))